src/person.cpp

```diff
@@ -1660,7 +1675,89 @@ AnimateResult Person::OnAnimate(int delay)
+                                               const TileEdge exit_edge,
+                                               const int dx,
+                                               const int dy,
+                                               const int dz)
+{
+	const XYZPoint16 init_pos = this->vox_pos;
+	Voxel *new_voxel = _world.GetCreateVoxel(this->vox_pos, true);
+	this->AddSelf(new_voxel);
+	this->UpdateZPosition();
+	new_voxel = _world.GetCreateVoxel(this->vox_pos, false);
+
+	assert(former_voxel->GetGroundType() != GTP_INVALID);
+	assert(new_voxel != nullptr && new_voxel->GetGroundType() != GTP_INVALID);
+	TileSlope old_voxel_slope = ExpandTileSlope(former_voxel->GetGroundSlope());
+	TileSlope new_voxel_slope = ExpandTileSlope(   new_voxel->GetGroundSlope());
+	/* Convert the bottom part of steep slopes to a format that is easier to handle here. */
+	auto convert_slope = [](TileSlope &slope) {
+		if ((slope & TSB_STEEP) == 0 || (slope & TSB_TOP) != 0) return;
+		const uint8 raised_corner_bit = (slope & ~TSB_STEEP);
+		slope = SL_FLAT;
+		if (raised_corner_bit != TSB_NORTH) slope |= TSB_SOUTH;
+		if (raised_corner_bit != TSB_SOUTH) slope |= TSB_NORTH;
+		if (raised_corner_bit != TSB_WEST ) slope |= TSB_EAST;
+		if (raised_corner_bit != TSB_EAST ) slope |= TSB_WEST;
+	};
+	convert_slope(old_voxel_slope);
+	convert_slope(new_voxel_slope);
+
+	uint16 left_height_1, left_height_2, right_height_1, right_height_2;
+	switch (exit_edge) {
+		case EDGE_NE:
+			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
+			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
+			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
+			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
+			break;
+		case EDGE_SE:
+			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
+			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
+			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
+			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
+			break;
+		case EDGE_SW:
+			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
+			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
+			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
+			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
+			break;
+		case EDGE_NW:
+			left_height_1  = former_vox_pos.z + ((old_voxel_slope & TSB_WEST ) == 0 ? 0 : 1);
+			right_height_1 = former_vox_pos.z + ((old_voxel_slope & TSB_NORTH) == 0 ? 0 : 1);
+			left_height_2  =  this->vox_pos.z + ((new_voxel_slope & TSB_SOUTH) == 0 ? 0 : 1);
+			right_height_2 =  this->vox_pos.z + ((new_voxel_slope & TSB_EAST ) == 0 ? 0 : 1);
+			break;
+		default: NOT_REACHED();
```

Comment:
If you first rotate the tile corners so the same edge is always used, these lines collapse in 1/4 of what you have now. `ROR` with 3 parameters is likely the simplest, it allows you to specify number of rotated bits. (Conceptually you rotate counter-clockwise, towards `EDGE_N` for the former voxel.)

I think the current voxel then also rotates counter-clockwise, so `ROR` would work as well.

As for adding, instead of testing for `0` and then adding 0 or 1, you can add the bit you test for after shifting to bit 0, eg
``` cpp
left_height_1  = former_vox_pos.z + ((old_voxel_slope >> TS_NORTH) & 1);  // Doesn't shift, as N is already bit 0.
right_height_1 = former_vox_pos.z + ((old_voxel_slope >> TS_EAST)  & 1);
```
I feel more is possible here, but not sure it's worth while (that may already hold for the test above, perhaps the compiler understand you test a bit and then add 1, it may do the conversion by itself already).

Reply:
I changed it to something even shorter (though slightly less readable), how about that